Make the code block header sticky

outdoor2kode opened the pull request:
This PR fixes #119.
Thought this feature will be really helpful if it can be implemented, ~plus it's a relatively simple tweak, so opened a PR for it.~

### Screencasts
#### WP_Query (Chrome)
https://user-images.githubusercontent.com/18050944/174574555-91505480-dec0-49d9-bff7-6d31018d5ff6.mp4
#### apply_filters (Firefox) 
https://user-images.githubusercontent.com/18050944/174574471-14968112-2475-4f54-87df-97b6fb1433b1.mp4

outdoor2kode committed:
Make the code block header sticky

outdoor2kode committed:
Make the header also sticky on a smaller screen

outdoor2kode commented on source/wp-content/themes/wporg-developer/scss/main.scss:
```diff
@@ -2388,6 +2388,7 @@ body.responsive-show {
 
 	#page {
 		overflow-x: hidden;
```
Since the `overflow-x: clip` is not supported on `Safari 15.5, 15.4` ([caniuse](https://caniuse.com/?search=overflow%3A%20clip)), and it's [suggested](https://github.com/WordPress/wporg-developer/pull/83#discussion_r891272205) to support them, so the attribute is retained for fallback purpose.

ryelle commented on source/wp-content/themes/wporg-developer/scss/prism.scss:
```diff
@@ -209,6 +209,9 @@ pre.line-numbers > code {
 /* Copy button */
 .wp-code-block-button-container {
+	position: sticky;
+	top: var(--wp-global-header-height);
```
This `--wp-global-header-height` value doesn't include the admin bar height, so if you're logged in the button is slightly hidden. You can use the `--wp-admin--admin-bar--height` custom property to include it.

```suggestion
	top: calc(var(--wp-global-header-height) + var(--wp-admin--admin-bar--height));
```

outdoor2kode replied:
Fixed in https://github.com/WordPress/wporg-developer/pull/125/commits/8c328d3dec4c625fdcea3ac34a0a0d7960cc360e

ryelle commented on source/wp-content/themes/wporg-developer/reference/template-source.php:
```diff
@@ -32,26 +32,25 @@
+			);
+		} ?>
 	</section>
+	<?php if ( ! empty( $source_code ) ) : ?>
+		<p class="source-code-links">
+			<span><a href="<?php echo get_source_file_link(); ?>"><?php _e( 'View on Trac', 'wporg' ); ?></a></span>
+			<span><a href="<?php echo get_github_source_file_link(); ?>"><?php _e( 'View on GitHub', 'wporg' ); ?></a></span>
+		</p>
+	<?php else : ?>
+		<p>
+			<a href="<?php echo get_source_file_link(); ?>"><?php _e( 'View on Trac', 'wporg' ); ?></a>
+		</p>
+	<?php endif; ?>
```
Why pull this out of the `section`?

outdoor2kode replied:
Adjusted in https://github.com/WordPress/wporg-developer/pull/125/commits/d247b95fb3b93757be23666280793b346190a5a3.

ryelle reviewed (commented):
I added an "explanation" locally (copied from [this page](https://developer.wordpress.org/reference/functions/add_menu_page/)) and the header here scrolls down out past the code:

![Screen Shot 2022-06-23 at 12 46 41](https://user-images.githubusercontent.com/541093/175352133-41e9199f-0909-45b5-ba8e-bc26d7226dfc.png)

outdoor2kode committed:
Rebase trunk

outdoor2kode committed:
Lint CSS

outdoor2kode committed:
Revert the pulling out of the section

Origianlly the section was pulled out for limiting the range the sticky
wp-code-block-button-container can scroll. But that way would make the
HTML structure a bit messy and not apply to all wp-block-code at once.
This new commit solves the problem.

outdoor2kode committed:
Support different screen size.

The wp-code-block-button-container will have expected layout
under different screen sizes.

outdoor2kode committed:
Make button container not overlap parent's border

outdoor2kode committed:
Fix sticky not work for shortcode issue

outdoor2kode commented on source/wp-content/themes/wporg-developer/scss/main.scss:
```diff
@@ -1328,11 +1328,6 @@
-
-			article {
-				overflow: auto;
-			}
```
`position: sticky` can't work correctly if its ancestors have `overflow: auto`.
See https://www.gatlin.io/blog/post/stuck-on-position-sticky
ps. Didn't spot any differences after removing this so far.

outdoor2kode commented on source/wp-content/themes/wporg-developer/scss/main.scss:
```diff
@@ -2397,7 +2392,8 @@ body.responsive-show {
 	#page {
-		overflow-x: hidden;
+		overflow-x: hidden; // a fallback for 'overflow: clip' on Safari
+		overflow-x: clip; // make 'position: sticky' work while preventing the overflow
```
`position: sticky` can't work correctly if its ancestors (other than `body` in our case) have `overflow: hidden`/`overflow: auto`/`overflow: scroll` as they would make these ancestors scrollable.
`overflow: clip` can fix the issue without losing the hidden effect.
See https://www.terluinwebdesign.nl/en/css/position-sticky-not-working-try-overflow-clip-not-overflow-hidden/.

outdoor2kode commented on source/wp-content/themes/wporg-developer/scss/prism.scss:
```diff
@@ -222,6 +222,10 @@ pre.line-numbers > code {
 	border-top-right-radius: 0.3em;
 
+	@media screen and (max-width: 876px) {
+		top: 0;
```
This is adjusted after rebasing the trunk. See https://github.com/WordPress/wporg-developer/pull/125/commits/f43a9e62989a1056179f0fb9011eae3ca1393ed1.

outdoor2kode commented:
### Updated Screencast

https://user-images.githubusercontent.com/18050944/191344970-a5d5e42b-80dd-451c-9f94-5d6022aceedf.mp4

outdoor2kode committed:
Rebase trunk

outdoor2kode committed:
Update packages

ryelle reviewed (approved):
Looks good! I checked a few pages with different code blocks and they all worked well 👍🏻
